src/instance/device/file.rs

```diff
@@ -0,0 +1,70 @@
+use async_trait::async_trait;
+use chrono::Utc;
+use std::time;
+use tokio::{fs::File, io::AsyncWriteExt};
+
+use crate::models;
+
+use super::{common::*, DeviceError};
+
+pub type FileDevice = Rewriter<FileDeviceImpl>;
+
+pub struct FileDeviceImpl {
+    leds: Vec<models::Color>,
+    print_timestamp: bool,
+    file_handle: File,
+    last_write_time: time::Instant,
+}
+
+#[async_trait]
+impl WritingDevice for FileDeviceImpl {
+    type Config = models::File;
+
+    fn new(config: &Self::Config) -> Result<Self, DeviceError> {
+        let file_handle = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .append(true)
+            .open(&config.output)?;
+
+        Ok(Self {
+            leds: vec![Default::default(); config.hardware_led_count as _],
+            print_timestamp: config.print_time_stamp,
+            file_handle: File::from_std(file_handle),
+            last_write_time: time::Instant::now(),
+        })
+    }
+
+    async fn set_let_data(
+        &mut self,
+        _config: &Self::Config,
+        led_data: &[models::Color],
+    ) -> Result<(), DeviceError> {
+        self.leds.copy_from_slice(led_data);
+        Ok(())
+    }
+
+    async fn write(&mut self) -> Result<(), DeviceError> {
+        let mut strbuf = String::new();
+        if self.print_timestamp {
+            // Prepend timestamp
+            let now = Utc::now();
+            let elapsed_time_ms = self.last_write_time.elapsed().as_millis();
+            self.last_write_time = time::Instant::now();
+
+            strbuf += &format!("{} | +{}", now, elapsed_time_ms);
+        }
+
+        // Join LED values into single string
+        let led_string = self.leds.iter().fold(String::new(), |a, b| {
+            a + &format!("{{{},{},{}}}", b.red, b.green, b.blue)
+        });
```

Comment:
fold and `+ &format` on strings will allocate at least N+1 times, use `write!` here too.

Reply:
Now using `.iter().for_each(...)` and then calling `write!`-macro in there.